File: frontend/src/domain/astrodynamics/Maneuvers.ts
/**
 * Orbital maneuver calculations.
 * References: Vallado §6.1-6.3, Curtis §6.1-6.3
 */

import { MU_EARTH } from './constants';
import { visViva, circularSpeed, orbitalPeriod } from './TwoBodyProblem';
import type { HohmannResult } from './types';

// ── Hohmann Transfer ─────────────────────────────────────────────────────────

/**
 * Hohmann transfer between two coplanar circular orbits.
 * Vallado §6.1
 *
 * @param r1_km  Initial circular orbit radius, km
 * @param r2_km  Final circular orbit radius, km
 * @param mu     Gravitational parameter, km³/s²
 */
export function hohmannTransfer(r1_km: number, r2_km: number, mu = MU_EARTH): HohmannResult {
  const aTransfer = (r1_km + r2_km) / 2;
  const eTransfer = Math.abs(r2_km - r1_km) / (r1_km + r2_km);

  // Velocity at periapsis of transfer ellipse
  const v_transfer_peri = visViva(r1_km, aTransfer, mu);
  // Velocity at apoapsis of transfer ellipse
  const v_transfer_apo  = visViva(r2_km, aTransfer, mu);

  const v1 = circularSpeed(r1_km, mu);
  const v2 = circularSpeed(r2_km, mu);

  const dv1 = v_transfer_peri - v1;  // prograde burn at r1
  const dv2 = v2 - v_transfer_apo;    // prograde burn at r2

  // Transfer time = half the period of the transfer ellipse
  const tof = orbitalPeriod(aTransfer, mu) / 2;

  return {
    dv1,
    dv2,
    dvTotal: Math.abs(dv1) + Math.abs(dv2),
    tof,
    aTransfer,
    eTransfer,
  };
}

// ── Bi-Elliptic Transfer ─────────────────────────────────────────────────────

/**
 * Bi-elliptic transfer (more efficient than Hohmann when r2/r1 > 11.94).
 * Three burns: r1 → rb (intermediate apoapsis) → r2
 *
 * @param r1_km  Initial orbit radius, km
 * @param rb_km  Intermediate apoapsis radius, km (r1 < rb, typically rb >> r1)
 * @param r2_km  Final orbit radius, km
 */
export function biEllipticTransfer(r1_km: number, rb_km: number, r2_km: number, mu = MU_EARTH) {
  const a1 = (r1_km + rb_km) / 2;  // first transfer ellipse
  const a2 = (rb_km + r2_km) / 2;  // second transfer ellipse

  const v1  = circularSpeed(r1_km, mu);
  const v2  = circularSpeed(r2_km, mu);
  const vb1 = visViva(r1_km, a1, mu);  // speed at r1 on first ellipse
  const va1 = visViva(rb_km, a1, mu);  // speed at rb on first ellipse (apo)
  const va2 = visViva(rb_km, a2, mu);  // speed at rb on second ellipse (peri)
  const vb2 = visViva(r2_km, a2, mu);  // speed at r2 on second ellipse

  const dv1 = vb1 - v1;               // first burn (prograde, at r1)
  const dv2 = Math.abs(va2 - va1);     // second burn (at rb)
  const dv3 = v2 - vb2;               // third burn (at r2)
  const dvTotal = Math.abs(dv1) + Math.abs(dv2) + Math.abs(dv3);

  const tof = orbitalPeriod(a1, mu) / 2 + orbitalPeriod(a2, mu) / 2;

  return { dv1, dv2, dv3, dvTotal, tof, a1Transfer: a1, a2Transfer: a2 };
}

// ── Plane Change ─────────────────────────────────────────────────────────────

/**
 * Pure plane change maneuver (no altitude change).
 * Δv = 2v·sin(Δi/2)
 *
 * @param v_km_s     Orbital speed at maneuver point, km/s
 * @param deltaI_rad Inclination change, radians
 */
export function planeChange(v_km_s: number, deltaI_rad: number): number {
  return 2 * v_km_s * Math.sin(deltaI_rad / 2);
}

/**
 * Combined plane change + altitude change (one burn, more efficient than sequential).
 * Δv = √(v1² + v2² - 2·v1·v2·cos(Δi))
 *
 * @param v1_km_s   Speed before maneuver, km/s
 * @param v2_km_s   Speed after maneuver, km/s
 * @param deltaI_rad Inclination change, radians
 */
export function combinedManeuver(v1_km_s: number, v2_km_s: number, deltaI_rad: number): number {
  return Math.sqrt(
    v1_km_s * v1_km_s
    + v2_km_s * v2_km_s
    - 2 * v1_km_s * v2_km_s * Math.cos(deltaI_rad)
  );
}

// ── Low-Thrust (Edelbaum) ────────────────────────────────────────────────────

/**
 * Edelbaum approximation for low-thrust orbit raising with inclination change.
 * More accurate than simple Δv for continuous thrust.
 *
 * @param v1_km_s    Initial circular orbit speed, km/s
 * @param v2_km_s    Final circular orbit speed, km/s
 * @param deltaI_rad Inclination change, radians
 */
export function edelbaumDeltaV(v1_km_s: number, v2_km_s: number, deltaI_rad: number): number {
  return Math.sqrt(
    v1_km_s * v1_km_s
    + v2_km_s * v2_km_s
    - 2 * v1_km_s * v2_km_s * Math.cos(Math.PI * deltaI_rad / 2)
  );
}

// ── Rendezvous (Clohessy-Wiltshire) ─────────────────────────────────────────

/**
 * CW (Hill's) equations for relative motion in LVLH frame.
 * Used for proximity operations and rendezvous.
 * Equations:
 *   ẍ - 2nẏ - 3n²x = fx  (radial, x positive outward)
 *   ÿ + 2nẋ = fy           (along-track, y positive prograde)
 *   z̈ + n²z = fz          (cross-track, z positive orbit-normal)
 *
 * @param n  Mean motion of target, rad/s
 */
export function cwPropagateRelative(
  state0: { x: number; y: number; z: number; dx: number; dy: number; dz: number },
  n: number,
  dt: number
): { x: number; y: number; z: number; dx: number; dy: number; dz: number } {
  const { x, y, z, dx, dy, dz } = state0;
  const s  = Math.sin(n * dt);
  const c  = Math.cos(n * dt);

  // Analytical CW solution (Clohessy & Wiltshire 1960)
  return {
    x:  (4 - 3 * c) * x + s / n * dx + 2 * (1 - c) / n * dy,
    y:  6 * (s - n * dt) * x + y - 2 * (1 - c) / n * dx + (4 * s - 3 * n * dt) / n * dy,
    z:  z * c + dz / n * s,
    dx: 3 * n * s * x + c * dx + 2 * s * dy,
    dy: -6 * n * (1 - c) * x - 2 * s * dx + (4 * c - 3) * dy,
    dz: -z * n * s + dz * c,
  };
}

// ── Delta-v Budget ───────────────────────────────────────────────────────────

/**
 * Tsiolkovsky rocket equation: Δv = Isp·g0·ln(m0/mf)
 * Solve for mass ratio given Δv and specific impulse.
 *
 * @param dv_km_s   Required Δv, km/s
 * @param isp_s     Specific impulse, seconds
 * @param g0_m_s2   Standard gravity, m/s² (default 9.80665)
 * @returns         Mass ratio m0/mf
 */
export function tsiolkovskyMassRatio(dv_km_s: number, isp_s: number, g0_m_s2 = 9.80665): number {
  return Math.exp(dv_km_s * 1000 / (isp_s * g0_m_s2));
}

/**
 * Propellant mass fraction from Tsiolkovsky equation.
 * @returns  Propellant fraction (0 to 1)
 */
export function propellantFraction(dv_km_s: number, isp_s: number): number {
  const mr = tsiolkovskyMassRatio(dv_km_s, isp_s);
  return 1 - 1 / mr;
}

File: backend/app/domain/astrodynamics/maneuvers.py
"""
Orbital maneuver calculations.
References: Vallado §6.1-6.3, Curtis §6.1-6.3
"""

import math
from .kepler import MU_EARTH


def circular_speed(r_km: float, mu: float = MU_EARTH) -> float:
    """Circular orbital speed at radius r_km [km/s]."""
    return math.sqrt(mu / r_km)


def vis_viva(r_km: float, a_km: float, mu: float = MU_EARTH) -> float:
    """Vis-viva equation: speed at radius r on orbit with SMA a [km/s]."""
    return math.sqrt(mu * (2 / r_km - 1 / a_km))


def orbital_period(a_km: float, mu: float = MU_EARTH) -> float:
    """Keplerian orbital period [seconds]."""
    return 2 * math.pi * math.sqrt(a_km**3 / mu)


def hohmann_transfer(r1_km: float, r2_km: float, mu: float = MU_EARTH) -> dict:
    """
    Hohmann transfer between two coplanar circular orbits.
    Vallado §6.1
    """
    a_t   = (r1_km + r2_km) / 2
    e_t   = abs(r2_km - r1_km) / (r1_km + r2_km)
    v1    = circular_speed(r1_km, mu)
    v2    = circular_speed(r2_km, mu)
    vt1   = vis_viva(r1_km, a_t, mu)
    vt2   = vis_viva(r2_km, a_t, mu)
    dv1   = vt1 - v1
    dv2   = v2 - vt2
    tof   = orbital_period(a_t, mu) / 2

    return {
        "dv1_km_s":      dv1,
        "dv2_km_s":      dv2,
        "dv_total_km_s": abs(dv1) + abs(dv2),
        "tof_s":         tof,
        "tof_min":       tof / 60,
        "a_transfer_km": a_t,
        "e_transfer":    e_t,
    }


def bielliptic_transfer(r1_km: float, rb_km: float, r2_km: float, mu: float = MU_EARTH) -> dict:
    """Bi-elliptic transfer. Three burns."""
    a1    = (r1_km + rb_km) / 2
    a2    = (rb_km + r2_km) / 2
    v1    = circular_speed(r1_km, mu)
    v2    = circular_speed(r2_km, mu)
    vb1   = vis_viva(r1_km, a1, mu)   # at r1 on first ellipse
    va1   = vis_viva(rb_km, a1, mu)   # at rb on first ellipse
    va2   = vis_viva(rb_km, a2, mu)   # at rb on second ellipse
    vb2   = vis_viva(r2_km, a2, mu)   # at r2 on second ellipse
    dv1   = vb1 - v1
    dv2   = abs(va2 - va1)
    dv3   = v2 - vb2
    tof   = orbital_period(a1, mu) / 2 + orbital_period(a2, mu) / 2

    return {
        "dv1_km_s":      dv1,
        "dv2_km_s":      dv2,
        "dv3_km_s":      dv3,
        "dv_total_km_s": abs(dv1) + abs(dv2) + abs(dv3),
        "tof_s":         tof,
        "a1_transfer_km": a1,
        "a2_transfer_km": a2,
    }


def plane_change(v_km_s: float, delta_i_rad: float) -> float:
    """Pure plane change: Δv = 2v·sin(Δi/2) [km/s]."""
    return 2 * v_km_s * math.sin(delta_i_rad / 2)


def combined_maneuver(v1_km_s: float, v2_km_s: float, delta_i_rad: float) -> float:
    """Combined altitude + plane change: Δv = √(v1² + v2² − 2v1v2cos(Δi))."""
    return math.sqrt(
        v1_km_s**2 + v2_km_s**2 - 2 * v1_km_s * v2_km_s * math.cos(delta_i_rad)
    )


def edelbaum_delta_v(v1_km_s: float, v2_km_s: float, delta_i_rad: float) -> float:
    """Edelbaum low-thrust approximation."""
    return math.sqrt(
        v1_km_s**2 + v2_km_s**2 - 2 * v1_km_s * v2_km_s * math.cos(math.pi * delta_i_rad / 2)
    )


def tsiolkovsky_mass_ratio(dv_km_s: float, isp_s: float, g0: float = 9.80665) -> float:
    """Tsiolkovsky rocket equation: mass ratio m0/mf."""
    return math.exp(dv_km_s * 1000 / (isp_s * g0))
